package/MDAnalysis/analysis/utils/autocorrelation.py

```diff
@@ -0,0 +1,155 @@
+# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
+# vim: tabstop=4 expandtab shiftwidth=4 softtabstop=4
+#
+# MDAnalysis --- https://www.mdanalysis.org
+# Copyright (c) 2006-2017 The MDAnalysis Development Team and contributors
+# (see the file AUTHORS for the full list of names)
+#
+# Released under the GNU Public Licence, v2 or any higher version
+#
+# Please cite your use of MDAnalysis in published work:
+#
+# R. J. Gowers, M. Linke, J. Barnoud, T. J. E. Reddy, M. N. Melo, S. L. Seyler,
+# D. L. Dotson, J. Domanski, S. Buchoux, I. M. Kenney, and O. Beckstein.
+# MDAnalysis: A Python package for the rapid analysis of molecular dynamics
+# simulations. In S. Benthall and S. Rostrup editors, Proceedings of the 15th
+# Python in Science Conference, pages 102-109, Austin, TX, 2016. SciPy.
+# doi: 10.25080/majora-629e541a-00e
+#
+# N. Michaud-Agrawal, E. J. Denning, T. B. Woolf, and O. Beckstein.
+# MDAnalysis: A Toolkit for the Analysis of Molecular Dynamics Simulations.
+# J. Comput. Chem. 32 (2011), 2319--2327, doi:10.1002/jcc.21787
+#
+
+from __future__ import absolute_import, division
+import numpy as np
+from copy import deepcopy
+
+
+def autocorrelation(list_of_sets, tau_max, window_step=1):
+    r"""The descrete implementation of the autocorrelation function.
+
+    .. math::
+       C_{HB}^c(\tau) = \frac{\sum_{ij}h_{ij}(t_0)h'_{ij}(t_0+\tau)}{\sum_{ij}h_{ij}(t_0)}
+
+    Parameters
+    ----------
+    list_of_sets : list
+      List of sets,
+    tau_max : int
+      The last tau (inclusive) for which to carry out autocorrelation.
+    window_step : int, optional
+      The step for the t0 to perform autocorrelation (without the overlap). Default is 1.
+
+    Returns
+    --------
+    tau_timeseries : list of int
+        the tau for which the autocorrelation was calculated
+    timeseries : list of int
+        the autocorelation values for each of the tau values
+    timeseries_data : list of list of int
+        the raw data from which the autocorrelation is computed. The time dependant evolution can be investigated.
+
+    .. versionadded:: 0.19.2
+    """
+
+    # Check types
+    assert type(list_of_sets) == list, "list_of_sets must be a one-dimensional list of sets"
```

Comment:
Please raise exceptions instead of using `assert` for input checking. This looks like `TypeError`.

(Test for exceptions are very welcome but if you really don't have time, mark them explicitly as not covered ... there's a pragma for coverage to tag blocks or lines of code as explicitly not tested. At least then we know where to send people for more GSoC issues...)

`assert` can be optimized away and it makes for terrible user experience.